### PR TITLE
K8s/Folders: Add folders api service (with legacy storage)

### PR DESCRIPTION
**Pull request by ryantxu**
Removes the cue folder definition, and replaces it with a basic api server talking to legacy storage.

This requires dev mode and:
```
[feature_toggles]
grafanaAPIServer = true
grafanaAPIServerEnsureKubectlAccess = true ; avoid the self signed SSL
grafanaAPIServerWithExperimentalAPIs = true
```

image

### CODE REVIEW

File: pkg/infra/appcontext/user.go
Comment (ryantxu) [inline pkg/infra/appcontext/user.go:63]: I am still getting:
```
kubectl create -f pkg/tests/apis/folders/testdata/folder-generate.yaml
Error from server: error when creating "pkg/tests/apis/folders/testdata/folder-generate.yaml": access denied to folder
```
but this fixed the list issues, and seems reasonable as a fallback for the super user

File: public/app/core/services/backend_srv.ts
Comment (ryantxu) [inline public/app/core/services/backend_srv.ts:524]: we could remove these deprecations, but enough tests have mocked responses i think we can do that in an additional pass

File: pkg/registry/apis/folders/legacy_storage.go
Comment (ryantxu) [reply pkg/registry/apis/folders/legacy_storage.go:89]: > without setting ParentUID returns the root folders

for this response we need *all* folders -- regardless where they live.  Likely we can add that to folder service

File: pkg/infra/appcontext/user.go
Comment (ryantxu) [reply pkg/infra/appcontext/user.go:63]: I think *:* is fine for this case -- this is supposed to be the user that is allowed to do anything everywhere

File: pkg/registry/apis/folders/legacy_storage.go
Comment (papagian) [reply pkg/registry/apis/folders/legacy_storage.go:89]: if `nestedFolders` are not enabled then all the folders are root folders (it's not possible to create subfolders)

**Comment by ryantxu** [comment]
This requires a dev feature flag to run, so I think it is safe to move forward.

The "Possible breaking changes" is from removing the unused cue definition and related typescript (also unused/invalid)

File: pkg/registry/apis/folders/register.go
Comment (ryantxu) [inline pkg/registry/apis/folders/register.go:136]: we could punt on these for now since nested folders won't be enabled -- but also they do not hurt and i think/hope make the long term strategy more clear

File: pkg/registry/apis/folders/sub_children.go
Comment (DanCech) [inline pkg/registry/apis/folders/sub_children.go:51]:
```suggestion
		children, err := r.service.GetChildren(ctx, &folder.GetChildrenQuery{
```

File: pkg/registry/apis/apis.go
Comment (DanCech) [inline pkg/registry/apis/apis.go:8]: we should set a standard for the group names, right now we have playlist (singular) and folders (plural)...

File: pkg/infra/appcontext/user.go
Comment (DanCech) [reply pkg/infra/appcontext/user.go:63]: I suspect that the problem is that folders does its own access checks and your signedinuser doesn't match those. Using the kubeconfig from the storage readme does work since you have a legit user

File: pkg/registry/apis/apis.go
Comment (ryantxu) [reply pkg/registry/apis/apis.go:8]: 🤔 any preference?  I flip back and forth, and in general looks like there is not a fixed convention.  If it is a single resource, the plural will be in the URL.

Comment (ryantxu) [reply pkg/registry/apis/apis.go:8]: plural feels natural for many, but not all -- i don't have any strong convictions here :)

Comment (DanCech) [reply pkg/registry/apis/apis.go:8]: for this case I think plurals make more sense, playlists.grafana.app feels more natural than folder.grafana.app

File: pkg/registry/apis/folders/legacy_storage.go
Comment (papagian) [inline pkg/registry/apis/folders/legacy_storage.go:92]:
```suggestion
		ist.Continue = base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("%d", page+1)))
```

I have tested the pagination and it works if I add the following block:
```
	page := 1
	if options.Continue != "" {
		continueVal, err := base64.StdEncoding.DecodeString(options.Continue)
		if err != nil {
			return nil, err
		}
		page, err = strconv.Atoi(string(continueVal))
		if err != nil {
			return nil, err
		}
	}
```

and then pass the `page` to the `GetChildren()`

File: pkg/registry/apis/folders/legacy_storage.go
Comment (papagian) [inline pkg/registry/apis/folders/legacy_storage.go:60]: Is it possible to filter by the `folder`? I'm trying using `fieldSelector` but it seems that it needs some additional functionality that we don't support?

File: pkg/registry/apis/folders/legacy_storage.go
Comment (ryantxu) [reply pkg/registry/apis/folders/legacy_storage.go:76]: I'll replace it with an error if nested folders are enabled 🤔

File: pkg/registry/apis/folders/legacy_storage.go
Comment (ryantxu) [reply pkg/registry/apis/folders/legacy_storage.go:62]: good question -- this is still TODO.  (not this PR)

**Comment by ryantxu** [comment]
@papagian -- can you take another look?  I moved the continue token stuff into a helper

File: pkg/registry/apis/folders/legacy_storage.go
Comment (papagian) [inline pkg/registry/apis/folders/legacy_storage.go:261]: If the folder contains alert rules then the request will fail unless we set `ForceDeleteRules` to `true`. But I think it's safer not to enable it.

File: pkg/registry/apis/folders/legacy_storage.go
Comment (papagian) [inline pkg/registry/apis/folders/legacy_storage.go:223]:
```suggestion
		OrgID:        info.OrgID,
		Overwrite:    true,
```
otherwise, it fails with:
```
for: "pkg/tests/apis/folders/testdata/folder-generate.yaml": error when patching "pkg/tests/apis/folders/testdata/folder-generate.yaml": the folder has been changed by someone else
```
In a follow up we could conditionally set it but I think it's ok to set it to true for now.

I also get the following warning:
```
Warning: resource folders/ccb7f960-a426-4b9a-a91f-04da6d7df12a is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.
```

**Review by papagian — COMMENTED** [review]
**Update: This is an unified-storage issue not related (blocking) with the specific PR**

I have tried creating a subfolder (by enabling the `nestedFolders` feature) using:
```
apiVersion: folders.grafana.app/v0alpha1
kind: Folder
metadata:
  generateName: x # anything is ok here... except yes or true -- they become boolean!
  annotations: 
    grafana.app/folder: f6f9b702-2162-47e8-8d20-cf1441b44ed2
spec:
  title: subfolder
  description: A subfolder
```
 and even though the folder is created successfully:
```
$ kubectl --kubeconfig=./grafana.kubeconfig create -f pkg/tests/apis/folders/testdata/folder-generate.yaml 
folder.folders.grafana.app/ed927380-5410-4a40-8704-efdd595042e8 created
```

 and the parent information is stored in the legacy store:
```
mysql> select * from folder where uid = 'ed927380-5410-4a40-8704-efdd595042e8';
+--------+--------------------------------------+--------+-----------+-------------+--------------------------------------+---------------------+---------------------+
| id     | uid                                  | org_id | title     | description | parent_uid                           | created             | updated             |
+--------+--------------------------------------+--------+-----------+-------------+--------------------------------------+---------------------+---------------------+
| 589834 | ed927380-5410-4a40-8704-efdd595042e8 |      1 | subfolder | A subfolder | f6f9b702-2162-47e8-8d20-cf1441b44ed2 | 2023-12-19 12:12:05 | 2023-12-19 12:12:05 |
+--------+--------------------------------------+--------+-----------+-------------+--------------------------------------+---------------------+---------------------+
1 row in set (0.01 sec)
```

it's missing from the `entity_folder` table (also there is no respective entry in the `entity_folder` table):
```
mysql> select * from entity where name = 'ed927380-5410-4a40-8704-efdd595042e8';
+--------------------------------------+---------------------+---------------------------------------------------------------------------+-----------+---------------------+---------------+----------+--------------------------------------+--------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------+--------+------+----------------------------------+---------------+--------------+---------------+------------+--------+------------+-----------+-------+------+-------------+---------+--------+--------+--------+
| guid                                 | version             | key                                                                       | namespace | group               | group_version | resource | name                                 | folder | meta                                                                                                                                                                                                                                                                                                              | body                                              | status | size | etag                             | created_at    | created_by   | updated_at    | updated_by | origin | origin_key | origin_ts | title | slug | description | message | labels | fields | errors |
+--------------------------------------+---------------------+---------------------------------------------------------------------------+-----------+---------------------+---------------+----------+--------------------------------------+--------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------+--------+------+----------------------------------+---------------+--------------+---------------+------------+--------+------------+-----------+-------+------+-------------+---------+--------+--------+--------+
| 27033f9e-fb0f-4204-9ca5-37c139933ab9 | 1737083376366698496 | /folders.grafana.app/folders/default/ed927380-5410-4a40-8704-efdd595042e8 | default   | folders.grafana.app | v0alpha1      | folders  | ed927380-5410-4a40-8704-efdd595042e8 |        | {"metadata":{"name":"ed927380-5410-4a40-8704-efdd595042e8","namespace":"default","uid":"014093f0-691b-40c2-9387-91c375cffa78","creationTimestamp":"2023-12-19T12:12:05Z","annotations":{"grafana.app/originKey":"1611107","grafana.app/originName":"SQL","grafana.app/updatedTimestamp":"2023-12-19T10:12:05Z"}}} | {"title":"subfolder","description":"A subfolder"} | NULL   |   49 | 284b4d0c89ad02d898fe67655cae6669 | 1702987925036 | user:1:admin | 1702980725000 |            | SQL    | 1611107    |         0 |       |      |             |         | null   | null   | null   |
+--------------------------------------+---------------------+---------------------------------------------------------------------------+-----------+---------------------+---------------+----------+--------------------------------------+--------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------+--------+------+----------------------------------+---------------+--------------+---------------+------------+--------+------------+-----------+-------+------+-------------+---------+--------+--------+--------+
1 row in set (0.00 sec)
```

and also from the API response:
```
$ kubectl --kubeconfig=./grafana.kubeconfig describe folder ed927380-5410-4a40-8704-efdd595042e8 
Name:         ed927380-5410-4a40-8704-efdd595042e8
Namespace:    default
Labels:       <none>
Annotations:  grafana.app/createdBy: user:1:admin
              grafana.app/originKey: 1611107
              grafana.app/originName: SQL
              grafana.app/originTimestamp: 1970-01-01T00:00:00Z
              grafana.app/updatedTimestamp: 2023-12-19T10:12:05Z
API Version:  folders.grafana.app/v0alpha1
Kind:         Folder
Metadata:
  Creation Timestamp:  2023-12-19T12:12:05Z
  Resource Version:    1737083376366698496
  UID:                 27033f9e-fb0f-4204-9ca5-37c139933ab9
Spec:
  Description:  A subfolder
  Title:        subfolder
Events:         <none>
```

File: pkg/infra/appcontext/user.go
Comment (ryantxu) [reply pkg/infra/appcontext/user.go:64]: the wildcard does not work for folders because it uses the permissions exactly -- while other resources (dashboards) process wildcards.  I'll leave it and then add in explicit permissions where necessary

File: pkg/registry/apis/folders/legacy_storage.go
Comment (ryantxu) [reply pkg/registry/apis/folders/legacy_storage.go:223]: overwrite=true should be fine, we won't even get to this unless the revisionVersion is valid

The `kubectl.kubernetes.io/last-applied-configuration` warning is in the frontend -- because we are not saving their annotation.  The fallback behavior is to do server-side-apply, and that seems to work OK

File: pkg/registry/apis/apis.go
Comment (papagian) [reply pkg/registry/apis/apis.go:8]: I did some testing with the unified storage and I think that the [FolderGroupName](https://github.com/grafana/grafana/blob/52dc42545826b4bf51a4e1c28e11c30a134db9b2/pkg/services/store/entity/models.go#L10) should also change to plural (otherwise the folders are not stored in the [entity_folder](https://github.com/grafana/grafana/blob/52dc42545826b4bf51a4e1c28e11c30a134db9b2/pkg/services/store/entity/sqlstash/sql_storage_server.go#L474))

File: pkg/infra/appcontext/user.go
Comment (mildwonkey) [inline pkg/infra/appcontext/user.go:54]: i'm curious how will we get this to work with multiple orgs?

Comment (ryantxu) [reply pkg/infra/appcontext/user.go:54]: regular users have their orgId set on connection.  This hook is just for the special k8s privileged user that has access to everything.

For "normal" access every user has an org

Is the question, how can a user query across multiple orgs?  That is not supported by our existing APIs, so 🤷🏻 multi-tentnat/multi-org controllers will either need to wait for unified storage that will support that as a first class operation, or we will need to update our commands to support it.

File: pkg/registry/apis/apis.go
Comment (ryantxu) [reply pkg/registry/apis/apis.go:8]: I updated the keys here, and got it working enough that [updateFolderTree](https://github.com/grafana/grafana/blob/b5e19cf35fdbe6f8c95593a00e6ea903b06c9258/pkg/services/store/entity/sqlstash/folder_support.go#L37) is called and does. not crash, but clearly not working yet!

image

image

I'll take a look at that after this is merged... worried about growing the scope too much!

File: pkg/services/store/entity/migrations/entity_store_mig.go
Comment (ryantxu) [inline pkg/services/store/entity/migrations/entity_store_mig.go:131]: no need for title -- it can be joined with the GUID

File: pkg/services/store/entity/sqlstash/folder_support.go
Comment (ryantxu) [inline pkg/services/store/entity/sqlstash/folder_support.go:7]: @DanCech -- naming convention ok?

File: pkg/services/store/entity/sqlstash/folder_support.go
Comment (ryantxu) [inline pkg/services/store/entity/sqlstash/folder_support.go:45]: maybe the column should be "group_name" so we don't have to worry about escaping it?

Comment (papagian) [reply pkg/services/store/entity/sqlstash/folder_support.go]: I would suggest using [dialect.Quote()](https://github.com/grafana/grafana/blob/0424d44b39235c228ce0f2988d81a563dd1cf4ed/pkg/services/sqlstore/migrator/dialect.go#L19) to make sure that it works for all the backends.

**Review by papagian — APPROVED** [review]
I have tested and LGTM

**Pull request merged**
